text/0004-api-history-schema.md

```diff
@@ -0,0 +1,249 @@
+#### `win.setTrafficLightPosition(position)` _macOS_
+
+* `position` [Point](structures/point.md)
+
+Set a custom position for the traffic light buttons. Can only be used with `titleBarStyle` set to `hidden`.
+```
+
+After this RFC, they will consult the schema, and also add an API history block:
+
+````markdown
+#### `win.setTrafficLightPosition(position)` _macOS_
+
+<!--
+```YAML history
+added:
+  - pr-url: https://github.com/electron/electron/pull/22533
+```
+-->
+
+* `position` [Point](structures/point.md)
+
+Set a custom position for the traffic light buttons. Can only be used with
+`titleBarStyle` set to `hidden`.
+````
+
+Which, after several PRs, may reach this state:
+
+```yaml
+added:
+  - pr-url: https://github.com/electron/electron/pull/22533
+changes:
+  - pr-url: https://github.com/electron/electron/pull/26789
+    description: Made `trafficLightPosition` option work for `customButtonOnHover` window.
+deprecated:
+  - pr-url: https://github.com/electron/electron/pull/37094
+    breaking-changes-header: deprecated-browserwindowsettrafficlightpositionposition
+```
+
+> [!NOTE]
+>
+> - *"Why no version numbers?"*
+>   - They will be derived from the PRs. This removes the need to change API History
+> on backports.
+>
+> - *"Why no `removed` property?"*
+>   - When an API is removed from Electron, it is also removed from the documentation.
+>
+> For further details, see [**Rationale and alternatives**](#rationale-and-alternatives).
+
+Meanwhile, the Electron website displays this information nicely in a table.
+Users can then use this information to help them upgrade their Electron version,
+migrate from deprecated APIs, etc.
+
+The schema provides information to the developer about what to include in the
+block, the minimum/maximum length of text, the required properties, etc.
+
+Documentation for older APIs will also eventually be changed to include these
+API history blocks.
+
+## Reference-level explanation
+
+The JSON Schema looks like this:
+
+```json
+{
+  "title": "JSON schema for API history blocks in Electron documentation",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "$comment": "If you change this schema, remember to edit the TypeScript interfaces in the linting script.",
+  "definitions": {
+    "changeSchema": {
+      "type": "object",
+      "properties": {
+        "pr-url": { "type": "string", "pattern": "^https://github.com/electron/electron/pull/\\d+$" },
+        "breaking-changes-header": { "type": "string", "minLength": 3 },
+        "description": { "type": "string", "minLength": 3, "maxLength": 72 }
+      },
+      "required": [ "pr-url" ],
+      "additionalProperties": false
+    }
+  },
+  "type": "object",
+  "properties": {
+    "added": { "type": "array", "items": { "$ref": "#/definitions/changeSchema" } },
+    "deprecated": { "type": "array", "items": { "$ref": "#/definitions/changeSchema" } },
+    "changes": { "type": "array", "items": { "$ref": "#/definitions/changeSchema" } }
```

Comment:
I think this would be better (for linting and validation reasons) if we didn't reuse `#/definitions/changeSchema` for each of these. You could still use definitions to avoid duplicating the property definitions.

Reason being, I think we want to enforce some further constraints, like `changes` should require `description`. And arguably `deprecated` should require `breaking-changes-header`.

Reply:
How's this?

```patch
diff --git a/text/0004-api-history-schema.md b/text/0004-api-history-schema.md
index 28d9075..f8e2590 100644
--- a/text/0004-api-history-schema.md
+++ b/text/0004-api-history-schema.md
@@ -108,7 +108,7 @@ The JSON Schema looks like this:
   "$schema": "http://json-schema.org/draft-07/schema#",
   "$comment": "If you change this schema, remember to edit the TypeScript interfaces in the linting script.",
   "definitions": {
-    "changeSchema": {
+    "baseChangeSchema": {
       "type": "object",
       "properties": {
         "pr-url": { "type": "string", "pattern": "^https://github.com/electron/electron/pull/\\d+$" },
@@ -117,13 +117,22 @@ The JSON Schema looks like this:
       },
       "required": [ "pr-url" ],
       "additionalProperties": false
+    },
+    "addedChangeSchema": {
+      "allOf": [ { "$ref": "#/definitions/baseChangeSchema" } ]
+    },
+    "deprecatedChangeSchema": {
+      "allOf": [ { "$ref": "#/definitions/baseChangeSchema" }, { "required": [ "breaking-changes-header" ] } ]
+    },
+    "changesChangeSchema": {
+      "allOf": [ { "$ref": "#/definitions/baseChangeSchema" }, { "required": [ "description" ] } ]
     }
   },
   "type": "object",
   "properties": {
-    "added": { "type": "array", "items": { "$ref": "#/definitions/changeSchema" } },
-    "deprecated": { "type": "array", "items": { "$ref": "#/definitions/changeSchema" } },
-    "changes": { "type": "array", "items": { "$ref": "#/definitions/changeSchema" } }
+    "added": { "type": "array", "items": { "$ref": "#/definitions/addedChangeSchema" } },
+    "deprecated": { "type": "array", "items": { "$ref": "#/definitions/deprecatedChangeSchema" } },
+    "changes": { "type": "array", "items": { "$ref": "#/definitions/changesChangeSchema" } }
   },
   "additionalProperties": false
 }

```

Reply:
That looks good, although I wouldn't make `breaking-changes-header` required just because there may be some (unfortunate) cases in historical data where it's missing. Since it's just an enhancement so we can link to it, I think it's fine to keep it optional.

Reply:
Resolved in [`24a4a55` (#6)](https://github.com/electron/rfcs/pull/6/commits/24a4a55dd4c5bb4364103610ecd396e35e4b1b66)